Updated the template, and added "hosted by Leafcloud"

dhrp opened the pull request:
As discussed, and as presented here is the PR with the template update.

I kept the changes as much as possible to the template only, but separated css and js with go templates for readability. 

Feedback welcome.

There are a few quirk fixes / improvements that would be cool for a later time.

Todo: 
- [x] Mobile view
- [x] Fix issue with width being too wide (= grid issue)
- [x] Fixed layout (width)
- [x] Something about which button (tab) is active 
- [x] Put logo behind flag
- [x] Less white input fields on dark mode
- [x] HIde Geolite mention when not used

~Something else when a field is selected that is undefined~
~Re-introduce alternative clients for curl ?~

maximbaz commented:
Given that many people selfhost this service, how about hiding the Leafcloud banner in the top right corner behind a CLI flag? Considering that we only want to show the banner in one specific deployment (on ifconfig.co) and not show it on any other deployment, I would make it disabled by default. Appreciate you hosting it for us on ifconfig.co by the way!

Some feedback, hope you find it useful 🙂 :

- The port checkbox is confusing to me, it is just another tab like "ip", "country", etc, but it's in a completely different location - I would move "port" to those tabs, just when you click on "port" another field appears somewhere nearby (next to "port", or above "curl example").
- There is no indication which of the tabs is an active one, i.e. what is the curl example for - I'd replace the buttons with something that resembles tabs a bit more, with an indication of the currently active one.
- On my screen I have a horizontal bar and the textareas in the right column still don't fit the screen:

    ![image](https://user-images.githubusercontent.com/1177900/101216810-76b1a700-3680-11eb-8281-cf418d236dc3.png)

- The choice of columns on the left side is a bit weird, I doubt "IP decimal" is really useful to be shown there, and also splitting User Agent into three separate rows... Honestly I'd just keep "IP address" and "User Agent" (with the raw full value).
- The note about GeoLite databases should be hidden when they are not used, like in my example above.

dhrp commented:
Thanks for the feedback @maximbaz 

I just fixed the horizontal layout issue, that should be fixed now. I've added most of your other points to my to-do list, and will react to some others here..

> The port checkbox is confusing to me, it is just another tab like "ip", 

Well, not exactly this information is not available by default / by example. It needs to be checked. When you press the button the front-end will actually fetch that data from the current site and should be able to display it without a page refresh. So if you would want it over the top it would also need a button there. 

I personally think a port checker would a nice feature. -- Though I actually have a question about this feature, because it was not working for me; even when the -r flag is specified.

> On my screen I have a horizontal bar and the textareas in the right column still don't fit the screen

The 'doesn't fit' should be fixed. The horizontal scrollbar in the outputbox is to accomodate the JSON response and is there because overflow-x: scoll is set. Windows just puts the scrollbar always. It's not really something I can change.

> The choice of columns on the left side is a bit weird, I doubt "IP decimal" is really useful to be shown there, and also splitting User Agent into three separate rows... Honestly I'd just keep "IP address" and "User Agent" (with the raw full value).

I would actually like to see user-agent (no raw value) as an endpoint. I think that would be useful. -- And because you're checking localhost you see very little fields, perhaps it makes more sense to see this once you see a long table, when fetching another IP? I don't really care which fields are shown, I just put all available ones. @mpolden: what is your opinion?

mpolden commented:
> how about hiding the Leafcloud banner in the top right corner behind a CLI flag? 

We could just check for ifconfig.co host header in the template, and hide it for other hosts.

> Well, not exactly this information is not available by default / by example. It needs to be checked. When you press the button the front-end will actually fetch that data from the current site and should be able to display it without a page refresh. So if you would want it over the top it would also need a button there.

True, but it's just another path. All paths except `/port` accept an optional `?ip=:ip` parameter, while port requires a port to check: `/port/:port`.

Maybe we can have an input field below the tabs, which takes an IP if any tab that's not "port" is selected, and a port for the "port" tab?

Something like this, where [thing] means selected:
```
ip country country-iso [city] asn port
Enter IP to check (optional): 1.1.1.1 (button that is enabled if IP is entered)
curl localhost:8080/city?ip=1.1.1.1
```

```
ip country country-iso city asn [port]
Enter a port to check (required): 22 (button)
curl localhost:8080/port/22
```

dhrp commented:
ok. I've updated the PR, should be good now.

- I've implemented showing the banner (or not) with a flag, instead of by hostname, because it's otherwise very hard to preview.
- I've implemented the port thing as requested, there is now no longer any fetch logic.

mpolden commented on Dockerfile.geoip:
```diff
@@ -0,0 +1,18 @@
+# Build
```
Can you remove this file? One `Dockerfile` is enough, and the Maxmind guys are very picky about redistribution/attribution of the database files.

mpolden commented on html/index.html:
```diff
@@ -0,0 +1,332 @@
+                  class="button widget-select"
+                  onclick="changeInput(this.name, this)"
+                >
+                  ASN
```
Can you make this lowercase so it matches the path name exactly?

mpolden commented on html/index.html:
```diff
@@ -0,0 +1,332 @@
+                  class="button widget-select"
+                  onclick="changeInput(this.name, this)"
+                >
+                  JSON
```
Same as above.

mpolden reviewed (commented):
Thanks! I added some minor comments, but this is looking ready for merge. What do you think, @maximbaz?

mpolden commented:
The port selector looks a bit off in Safari, but not critical to fix. 

<img width="500" alt="Screen Shot 2020-12-13 at 22 14 54" src="https://user-images.githubusercontent.com/174496/102024213-04804700-3d91-11eb-97ab-f7b977676ec8.png">

maximbaz commented on html/index.html:
```diff
@@ -0,0 +1,332 @@
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <meta
+      name="description"
+      content="ifconfig. What is my IP address? &mdash; The best tool to find your own IP address, and information about it."
```
Should `ifconfig.` be removed, or replaced with `{{ .Host }}` perhaps?

maximbaz commented on html/index.html:
```diff
@@ -0,0 +1,332 @@
+      name="description"
+      content="ifconfig. What is my IP address? &mdash; The best tool to find your own IP address, and information about it."
+    />
+    <link rel="canonical" href="https://ifconfig.co/" />
```
Should be replaced with `{{ .Host }}`, for selfhosted use-case?

dhrp replied:
This should point to the one only ifconfig.co. See https://developers.google.com/search/docs/advanced/crawling/consolidate-duplicate-urls

Otherwise, if people use a google search google may have many pages which look exactly the same, but hosted on different domains. It may then decide to distribute the search ranking over all of them. -- making ifconfig.co less findable. 

This tag tells search engines that ifconfig.co is the "canonical" URL

maximbaz commented on html/index.html:
```diff
@@ -0,0 +1,332 @@
+                  <a href="https://www.leaf.cloud?utm_source=ifconfig"
+                    >Leafcloud – The Truly Sustainable Cloud</a
+                  >
+                </p>
```
not sure if html formatting like above is intentional? 🤔

dhrp replied:
√

maximbaz commented on html/index.html:
```diff
@@ -0,0 +1,332 @@
+              MaxMind, available from
+              <a href="https://www.maxmind.com">www.maxmind.com</a>
+            </p>
+            {{ end }} {{ if .City }}
```
Technically you need to check latitude and longitude instead of city, I guess, unless you know that those parameters always exist if city is known?

dhrp replied:
√

maximbaz reviewed (approved):
Good job, some very minor comments 👍

dhrp commented:
<img width="552" alt="Screen Shot 2020-12-14 at 12 41 07" src="https://user-images.githubusercontent.com/2017300/102077493-ae9bb580-3e09-11eb-99cb-2de854f459ee.png">

mpolden commented:
Thanks, @dhrp!

Fossil01 commented:
IPv6 addresses are cutoff now with the new UI: 
![image](https://user-images.githubusercontent.com/8832687/103213726-dc572300-490e-11eb-8e65-a3d1ac2c2bce.png)

dhrp commented:
I'm going to submit a fix of this ipv6 issue and one other soon.

Fossil01 commented:
@dhrp Was this ever done? My IPv6 addresses still don't fit on a single line.
